compair/tests/algorithms/test_pair_adaptive.py

```diff
@@ -212,12 +212,86 @@ def test_generate_pair(self, mock_shuffle):
             )
         ]
         comparisons = [
-            ComparisonPair(1,2,None)
+            ComparisonPair(1,2,None),
+            ComparisonPair(3,4,None)
         ]
 
         results = self.pair_algorithm.generate_pair(scored_objects, comparisons)
 
         # object 1 should be selected since random shuffle is disabled
         self.assertEqual(results.key1, 1)
         # object 4 should be selected as 0.4 is the closest value to 0.5
-        self.assertEqual(results.key2, 4)
\ No newline at end of file
+        self.assertEqual(results.key2, 4)
+
+        # ensure that all scored objects are seen only once until they have almost all been seen
+        # (even number of scored objects)
+        scored_objects = [ScoredObject(
+            key=index, score=None, variable1=None, variable2=None,
+            rounds=0, wins=None, loses=None, opponents=None
+        ) for index in range(30)]
+        comparisons = []
+
+        used_keys = set()
+        all_keys = set(range(30))
+
+        # n/2 comparisons = 15
+        for _ in range(15):
+            results = self.pair_algorithm.generate_pair(scored_objects, comparisons)
+            # neither key should have been seen before
+            self.assertNotIn(results.key1, used_keys)
+            self.assertNotIn(results.key2, used_keys)
+
+            comparisons.append(results)
+            used_keys.add(results.key1)
+            used_keys.add(results.key2)
+
+        self.assertEqual(used_keys, all_keys)
+
+        # remaining comparisons for n(n-1)/2 = 435
+        for _ in range(435 - 15):
+            results = self.pair_algorithm.generate_pair(scored_objects, comparisons)
+            # both keys should have been seen before
+            self.assertIn(results.key1, used_keys)
+            self.assertIn(results.key2, used_keys)
+            comparisons.append(results)
+
+        # next comparison should be an UserComparedAllObjectsException error
+        with self.assertRaises(UserComparedAllObjectsException):
+            results = self.pair_algorithm.generate_pair(scored_objects, comparisons)
+
```

Comment:
Maybe add checking that there is no duplication when the algorithm exhausted all possibilities
```
        # make sure all pairs are distinct
        self.assertEqual(
            len(comparisons), 
            len(set([tuple(sorted([c.key1, c.key2])) for c in comparisons])))
```
And similarly for odd number of scored obj (and other algo too)

Reply:
Great idea!